afed.1
.TH AFED 1 "January 12, 2023"
.SH NAME
afed - an interpreted functional language for interactive computations

.SH SYNOPSIS
.SY afed
[ \fB\-CdnE\fP ]
[ \fB\-e\fP \fI<ERR_FILE>\fP ]
[ \fB\-f\fP \fI<INPUT_PATH>\fP ]
[ \fB\-L\fP \fI<DIRECTORY>\fP ]
[ \fB\-\-no\-local\-pkgs\fP ]
[ \fI<INPUT>\fP [ \fI<OUTPUT>\fP ]]
.SY afed
[ \fB\-h\fP | \fB\-?\fP | \fB\-\-help\fP ]
.YS


.SH DESCRIPTION
Afed is an interpreted functional language which embraces a declarative style
for ASCII-based interactive computing.
As opposed to printing results to the command line or a file,
Afed places the results of computations into \fIequals expressions\fP in the document itself.
These \fIequals expressions\fP are indicated by '= ``'.
For example, this line
.IP
.EX
1 - (3 + 4 = ``) = ``
.EE
.P
after evaluation would become
.IP
.EX
1 - (3 + 4 = `7`) = `-6`
.EE
.P

The Afed programming language is primarily focused on supporting mathematical computations.
To this end, it has a number of libraries written in Rust that make it easy to solve mathematical problems.



.SH OPTIONS
\fBafed\fP GNU-like syntax for command line arguments.

.TP
.B <INPUT>
Afed program file that will be parsed and evaluated.
The results of each computation attached to an \fIequals expression\fP
will be printed in the area between the graves.
If parse errors occur nothing will be printed in that expression.
INPUT can be '-' to pull the file from STDIN.
If not given the input defaults to STDIN.

.TP
.B <OUTPUT>
Filename to which the results will be written.
If there is no output given then the result is written back to the input.
OUTPUT can be '-' to print to STDOUT or '-2' to print to STDERR.

.TP
.B \-C
.TQ
.B \-\-check
Parses and evaluates the document printing any errors encountered.
However, no results are sent to the output.

.TP
.B \-d
.TQ
.B \-\-clear
The text between the graves in each equals expression is deleted.
It makes no other modifications.

.TP
.B \-n
.TQ
.B \-\-no\-clobber
Ensures that the input file is not modified.
If \fB-n\fP is set and the input and output are the same
the program exit with an error.

.TP
.BI \-e\  <ERR_FILE>
.TQ
.BI \-\-errors\  <ERR_FILE>
Filename to which the parsing and evaluation errors will be written.
ERR_FILE can be '-' to write the errors to STDOUT.
If not given the program defaults to '-2' which is STDERR.

.TP
.B \-E
.TQ
.B \-\-no\-errors
Suppresses any parsing or evaluation errors from being written anywhere.
If present then any \fB-e\fP flag will be ignored.

.TP
.B \-f\  <INPUT_PATH>
.TQ
.B \-\-filename\  <INPUT_PATH>
Name used internally for INPUT.
This will show up when parsing and evaluation errors are printed.
This is most useful when you want to provide
the name of a file that is passed in through STDIN.

.TP
.B \-L\  <DIRECTORY>
.TQ
.B \-\-pkg\  <DIRECTORY>
Check for packages in <DIRECTORY> to load.
By default, \fBafed\fP will search the \fIprogram's config folder\fP for packages to load.
See the \fBPACKAGES\fP section.

.TP
.B \-\-no\-local\-pkgs
Prevent \fBafed\fP from searching for packages in the \fIconfig folder\fP.
Only directories given with \fB-L\fP will be searched.

.TP
.B \-h, \-?, \-\-help
Print a help message for the executable and exits.



.SH EXAMPLES
A typical call to Afed will parse, evaluate,
and then modify the input file.
.IP
.EX
afed input.af
.EE
.P

To designate a different place to send the output,
we can just add an argument.
.IP
.EX
afed input.af output.af
.EE
.P

To make sure, the output is different than the input,
we can use the \fB\-n\fP flag.
The following command will produce an error since a distinct output is not given.
.IP
.EX
afed \-\-no\-clobber input.af
.EE
.P

We could also just check for errors and remove any previous results with
.IP
.EX
afed input.af \-C \-d
.EE
.P

If we want to save the errors, we have
.IP
.EX
afed input.af --errors afed_errors.err
.EE
.P


.SH PACKAGES
Packages are collections of variables and functions under a given name.
They behave the same as maps in \fBafed\fP.
The builtin \fIpackages\fP are
.IP \fBmath\fP
Standard rounding, trigonometric, exponential, logarithmic, and combinatorics functions.
These are also accessible as methods on numbers.
For convenience, all members of this library are available in the global namespace.
.IP \fBarr\fP
Array manipulation functions including \fBmap\fP, \fBfilter\fP, \fBfold\fP, and others.
These functions can also be called as methods on arrays.
For convenience, all members of this library are available in the global namespace.
.P

Afed's standard functionality can be extended with dynamically linked \fIpackages\fP.
These are distributed as dynamically linked libraries (e.g. \fB.dll\fP or \fB.so\fP files).
By default \fBafed\fP will search for \fIpackages\fP in
\fI$XDG_CONFIG_HOME/afed/pkgs\fP or \fI$HOME/.config/afed/pkgs\fP on Linux systems,
\fI$HOME/Library/Application Support/Afed/pkgs\fP on macOS, and
\fI{FOLDERID_RoamingAppData}\\Afed\\config\\pkgs\fP on Windows.
The standard dynamically linked \fIpackages\fP are
.IP \fBnum\fP
Functions dealing with prime numbers and factorization of positive integers.
This includes \fBis_prime\fP and \fBprime_factors\fP.
A type for modular arithmetic constructed using \fBmod.Mod\fP
(\fBMod\fP is also available in the global namespace).
For example, \fI3 (mod 7)\fP is written '\fB3 * Mod 7\fP'.
.IP \fBlin\fP
Contains constructor for a vector type, \fBlin.V\fP, and a matrix type, \fBlin.M\fP.
Both of these constructors are available in the global namespace.
It also provides vector and matrix manipulations functions.
.IP \fBcalc\fP
Includes functions for calculating the derivative, integral, maximum, and minimum of real valued functions.
It also provides the a plot type with constructor, \fBcalc.Plot\fP for creating plots.
Plots can be used to created 2D graphs of real valued functions that are displayed in the document.
.P

To create an afed \fIpackage\fP, one can use the \fBafed_objects\fP crate.
This crate provides the basic \fBafed\fP datatypes as well as
several macros for making \fIpackages\fP.
Specifically, one must call the \fBdeclare_pkg\fP macro to indicate
which constants and functions should be exported.


.SH LANGUAGE
Afed supports a set of primitive types similar to JavaScript, Python, and many other interpreted languages.
This includes numerics, strings, booleans, a null type, arrays (lists), and maps (dictionaries).
Single line comments are started by '\fB#\fP' while multiline comments are delimited by '\fB#{\fP' and '\fB}#\fP'.
There are methods defined on many of these types which can be called like
.IP
.EX
3.log 81 = `4`,  # Log base 3 of 81
[1, 2, 3].len = `3`,
"foo".upper = `"FOO"`,
"Hello\\nWorld".print = `Hello
World`,
.EE
.P
Notice that \fB.print\fP can be used to print strings raw and unquoted.
Values can be assigned to variables and used later.
The order of declaration is irrelevant for evaluation.
These variables can be scoped using maps.
.IP
.EX
x: {
    y: 3 * u
},
u: 4,
x.y = `12`,
.EE
.P
Afed also has user-definable functions and lambda expressions.
At this time, recursive user-defined functions are not supported.
These are defined and called using a Haskell-like syntax.
.IP
.EX
f x y: x + y,
f 2 3 = `5`,
(\\x y: x + y) 2 3 = `5`,
.EE
.P
These definitions also support destructuring of arrays and maps.
For maps, the destructuring may be fuzzy about the presence of other keys.
.IP
.EX
f [_, x, y, _] {a: c, ..}: x * y + c,
f [0, 3, 2, -1] {a: 5, b: ""} = `11`,
.EE
.P
The '\fB_\fP' indicates an argument that will be ignored. It may be present zero or more times.
The '\fB..\fP' indicates that the map may have unmatched keys.
Destructuring can also be used to define variables.
.IP
.EX
[{q: [_, y], r: z}, _, x]: [{"q": [8, false], r: null}, 1, 2],
x = `2`, y = `false`, z = `null`,
.EE
.P

Classes can be defined by passing a map to the \fBClass\fP function.
The passed map contains the constructor (labelled "new"), the instance methods,
and the operator overloads (e.g. \fB__call\fP, \fB__add\fP, ...).
The string converter method is \fB__str\fP.
Like Python, the first argument of instance methods will be the instance itself.
The instance variables are called the \fIdata\fP.
It can be any object, though a map would be a prudent choice.
The \fIdata\fP for any instance will be set to the object returned by \fBnew\fP.
To construct instances inside of the methods, one can call \fB.new\fP on any existing instance.
.IP
.EX
Person: Class {
    POPULATION: 10 ^ 5,

    new n a: {name: n, age: a},
    __add self shift: self.new self.name (self.age + shift),

    mult_by_age self x: x * self.age,
    add_surname self lastname: self.new (self.name + " " + lastname) self.age,
    __str self: "Person(" + self.name + ", " + self.age.str + ")",
},
Person.POPULATION = `100000`,

george: Person "George" 35 = `Person(George, 35)`,
george.__data = `{age: 35, name: "George"}`,
george + 10 = `Person(George, 45)`,

george.mult_by_age 2 = `70`,
Person.mult_by_age george 3 = `105`,
george_jones: george.add_surname "Jones" = `Person(George Jones, 35)`,
.EE
.P
Here \fBPerson\fP serves as the constructor and representative of the class.
Whenever a attribute or method is accessed that is not defined in the class,
the \fIdata\fP will be queried instead.
Above the \fIdata\fP object is directly obtained using \fB__data\fP.
The constructor \fBnew\fP may be omitted.
Then, the constructor will take a single object and set it as the \fIdata\fP object.
\fBClass\fP may also be accessed as \fBcls.Class\fP.


More information about any objects, methods, or packages can be found by calling \fBhelp\fP.
For example, to get information about the \fBlen\fP method or the number type,
we can call
.IP
.EX
help 4 = ``,
help [1, true, 3].len = ``,
help arr = ``,
.EE
.P


.SH IMPORTING
An Afed file may import other Afed files with \fBuse\fP statements.
To import all of the top-level variable declarations from 'child.af' into 'parent.af'
(assuming 'child.af' and 'parent.af' are in the same directory),
we would place the following line into 'parent.af' outside of any map
.IP
.EX
use "child.af",
.EE
.P
We may also provide an absolute path to 'child.af'
or a path relative to where the \fBafed\fP command called.
We can also do qualified imports by encompassing the \fBuse\fP in a map
.IP
.EX
my_child: { use "child.af" },
.EE
.P
Note that \fBuse\fP can only occur inside of a map body.
Also, \fBuse\fP cannot be used as a variable name.

.SH AUTHOR
.IP
Tanner Reese <tannerjamesreese@gmail.com>



afed.c
#include "nmsp.h"
#include "docmt.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <errno.h>
#include <getopt.h>

#define PROG_NAME "afed"

// Location to read program from, write it to, and log errors
FILE *infile = NULL;
FILE *outfile = NULL;
FILE *errfile = NULL;

bool only_check = 0;
bool allow_overwrite = 1;
bool show_errors = 1;

const char help_msg[] =
	"Usage: " PROG_NAME " [OPTION]... [-i] INFILE [[-o] OUTFILE]\n"
	"\n"
	"Evaluate expressions in place\n"
	"\n"
	"  -i, --input INFILES...  List of files to evaluate\n"
	"  -o, --output OUTFILE    Output file to store result to\n"
	"  -C, --check             Don't output file only check for errors\n"
	"  -n, --no-clobber        Make sure none of the INFILES are used as outputs\n"
	"  -e, --errors ERRFILE    File to send errors to. Sent to stderr if not specified\n"
	"  -E, --no-errors         Don't print any error messages\n"
	"  -h, --help              Print this help message\n"
	"\n"
	"'-' may be used with -o, -i, or -e to indicate STDOUT, STDIN, or STDOUT, respectively\n"
;

const char usage_msg[] = "Usage: " PROG_NAME " [OPTION]... [-i] INFILE [-o OUTFILE]\nUse --help for more information";

struct option longopts[] = {
	{"input", required_argument, NULL, 'i'},
	{"output", required_argument, NULL, 'o'},
	{"check", no_argument, NULL, 'C'},
	{"no-clobber", no_argument, NULL, 'n'},
	{"errors", required_argument, NULL, 'e'},
	{"no-errors", required_argument, NULL, 'E'},
	{"help", no_argument, NULL, 'h'},
	{0}
};

void leave(int code){
	// Close open file descriptors
	if(infile) fclose(infile);
	if(outfile && outfile != infile) fclose(outfile);
	if(errfile) fclose(errfile);
	
	exit(code);
}

#define usage(code, ...) { fprintf(stderr, __VA_ARGS__); puts(usage_msg); leave(code); }

void parse_opt(int key){
	switch(key){
		case -1:  // Non-Option Arguments
			if(!infile){  // Check for already defined infile
		case 'i':  // Input file
				if(infile) usage(2, "Input file already given\n");
				if(optarg[0] == '-' && optarg[1] == '\0') infile = stdin;
				else infile = fopen(optarg, "r+");  // Allow modification in case infile is used as outfile
				
				// Check that it opened
				if(!infile) usage(1, "Input file \"%s\" did not open: ERRNO %i\n", optarg, errno);
			}else{
		case 'o':  // Output file
				if(outfile) usage(2, "Output file already given\n");
				if(optarg[0] == '-' && optarg[1] == '\0') outfile = stdout;
				else outfile = fopen(optarg, "w");
				
				// Check that it opened
				if(!outfile) usage(1, "Output file \"%s\" did not open: ERRNO %i\n", optarg, errno);
			}
		break;
		
		// Only check for errors
		case 'C': only_check = 1;
		break;
		
		// Don't allow overwriting of input files
		case 'n': allow_overwrite = 0;
		break;
		
		case 'e':  // Error file
			if(errfile) usage(4, "Error file already given\n");	
			if(optarg[0] == '-' && optarg[1] == '\0') errfile = stdout;
			else errfile = fopen(optarg, "w");
			
			// Check that it opened
			if(!errfile) usage(2, "Error File \"%s\" did not open: ERRNO %i\n", optarg, errno);
		break;
		
		// Don't show any errors
		case 'E': show_errors = 0;
		break;
		
		// Print help message
		case 'h':
			puts(help_msg);
			leave(0);
	}
}




// Read contents of file into heap allocated string
char *read_file(FILE *fl);

int main(int argc, char *argv[]){
	// Parse Command Line Arguments
	// -----------------------------
	int c;
	while((c = getopt_long(argc, argv, "i:o:e:nEC", longopts, NULL)) != -1) parse_opt(c);
	for(int i = optind; i < argc; i++){
		optarg = argv[i];
		parse_opt(-1);
	}
	
	// Make sure an infile was given
	if(!infile) usage(4, "No Input file given\n");
	
	// Set default outfile as infile unless --no-clobber present
	if(!outfile && !only_check){
		if(allow_overwrite) outfile = infile == stdin ? stdout : infile;
		else usage(3, "No Output file given and --no-clobber present\n");
	}
	
	// Set default error file if not given
	if(!errfile) errfile = stderr;
	
	// Get content of input file and process
	char *prog = read_file(infile);
	
	// Parse and evaluate
	namespace_t nmsp = nmsp_new(true);
	docmt_t doc = docmt_new(prog, nmsp);
	int errcnt = docmt_parse(doc, show_errors ? errfile : NULL);  // Parse file
	fseek(infile, 0, SEEK_SET);  // Move `infile` back to beginning
	
	// Print out to new file
	errcnt += docmt_fprint(doc,
		only_check ? NULL : outfile,
		show_errors ? errfile : NULL
	);
	
	if(only_check){
		// Print out number of errors
		if(errcnt > 0) fprintf(errfile, "%i Parse Error%c\n", errcnt, errcnt > 1 ? 's' : ' ');
		else fprintf(errfile, "No Parse Errors\n");
	}
	
	// Cleanup heap allocations
	free(prog);
	docmt_free(doc);
	nmsp_free(nmsp);
	leave(0);
}



char *read_file(FILE *fl){
	size_t len = 0, cap = 1024;  // Begin with 1024 bytes of capacity
	char *cont = malloc(sizeof(char) * cap);
	while((len += fread(cont + len, 1, cap - len, fl)) >= cap){  // Consume input until EOF
		cap <<= 1;
		cont = realloc(cont, sizeof(char) * cap);
	}
	return cont;
}

